core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -517,59 +530,62 @@ private void innerIndex(Index index) throws IOException {
             }
             final long expectedVersion = index.version();
             if (checkVersionConflict(index, currentVersion, expectedVersion, deleted)) {
-                index.setCreated(false);
-                return;
-            }
-            final long updatedVersion = updateVersion(index, currentVersion, expectedVersion);
-            index.setCreated(deleted);
-            if (currentVersion == Versions.NOT_FOUND && forceUpdateDocument == false) {
-                // document does not exists, we can optimize for create
-                index(index, indexWriter);
+                // skip index operation because of version conflict on recovery
+                return new IndexResult(expectedVersion, false, index.estimatedSizeInBytes());
```

Comment:
I changed it to have one return statement with a final variable and calling freeze on it as the last operation. thanks for pointing it out.

I agree we should remove the inconsistency in handling version conflict during recovery and for replicas. I will open a separate PR for it